### PR TITLE
boost: Enable Boost.Python (NumPy)

### PR DESCRIPTION
**Pull request by podsvirov**
Some libraries will added:

```console
2022-02-09T17:18:05.3661747Z  /mingw64/bin/libboost_math_tr1l-mt.dll
2022-02-09T17:18:05.3673859Z  /mingw64/bin/libboost_math_tr1-mt.dll
2022-02-09T17:18:05.3675776Z  /mingw64/bin/libboost_nowide-mt.dll
2022-02-09T17:18:05.3676521Z +/mingw64/bin/libboost_numpy39-mt.dll
2022-02-09T17:18:05.3677099Z  /mingw64/bin/libboost_prg_exec_monitor-mt.dll
2022-02-09T17:18:05.3677723Z  /mingw64/bin/libboost_program_options-mt.dll
2022-02-09T17:18:05.3678304Z  /mingw64/bin/libboost_python39-mt.dll
2022-02-09T17:18:05.3678775Z @@ -16350,6 +16351,8 @@
2022-02-09T17:18:05.3679653Z  /mingw64/lib/libboost_math_tr1-mt.dll.a
2022-02-09T17:18:05.3680233Z  /mingw64/lib/libboost_nowide-mt.a
2022-02-09T17:18:05.3680741Z  /mingw64/lib/libboost_nowide-mt.dll.a
2022-02-09T17:18:05.3681591Z +/mingw64/lib/libboost_numpy39-mt.a
2022-02-09T17:18:05.3682139Z +/mingw64/lib/libboost_numpy39-mt.dll.a
2022-02-09T17:18:05.3682779Z  /mingw64/lib/libboost_prg_exec_monitor-mt.a
2022-02-09T17:18:05.3683390Z  /mingw64/lib/libboost_prg_exec_monitor-mt.dll.a
2022-02-09T17:18:05.3684222Z  /mingw64/lib/libboost_program_options-mt.a
```

### CODE REVIEW

**Comment by Biswa96** [comment]
Try replacing g++ with `$(CXX)` in that shell script.

**Comment by MehdiChinoune** [comment]
the issue is on the start of the PKGBUILD: https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-boost/PKGBUILD#L3

**Comment by podsvirov** [comment]
For `MINGW32`, `MINGW64` and `UCRT64` all good, as expected.

**Comment by lazka** [comment]
hm, tricky, we explicitly don't set clang here, because the boost build system behaves different then.

**Comment by lazka** [comment]
uh, boost has all its mingw specific code in gcc specific logic, using their own DSL -> I give up

**Comment by MehdiChinoune** [comment]
Could someone report it to them!

**Comment by MehdiChinoune** [comment]
I think the solution was written just after the failure:
```
  For any toolset you can override the path to the compiler with the '--cxx'
  option. You can also use additional flags for the compiler with the
  '--cxxflags' option.
```

**Comment by lazka** [comment]
This is only for building the build-tool (b2), not for boost itself I think.

**Comment by MehdiChinoune** [comment]
> This is only for building the build-tool (b2), not for boost itself I think. But maybe I missed something.

Yes, I tried it, I didn't work.

**Comment by jeremyd2019** [comment]
I'm looking at this, think I've just about got it

**Comment by lazka** [comment]
(Otherwise we can do the "clang-as-gcc" split package thing for now)

**Comment by MehdiChinoune** [comment]
I think I found a trick:
```
./bootstrap.sh --with-icu=${MINGW_PREFIX} --with-python=${MINGW_PREFIX} --with-toolset="gcc --cxx=${MINGW_PREFIX}/bin/c++.exe"
```

**Comment by jeremyd2019** [comment]
I thought of that, but the project-config.jam would probably be messed up.

**Comment by jeremyd2019** [comment]
I pushed my idea here.

File: mingw-w64-boost/PKGBUILD [hunk 1]
```diff
@@ -168,9 +164,12 @@ build() {
   local _pyver=$(${MINGW_PREFIX}/bin/python -c "import sys;sys.stdout.write('.'.join(map(str, sys.version_info[:2])))")
 
   echo "using python : ${_pyver} : ${PREFIX_DEPS}/bin/python.exe : ${PREFIX_DEPS}/include/python${_pyver} : ${PREFIX_DEPS}/lib/python${_pyver}/config-${_pyver} ;" >> ./tools/build/src/user-config.jam
-  #echo "using gcc : : ${PREFIX_DEPS}/bin/g++.exe : <rc>${PREFIX_DEPS}/bin/windres.exe ;" >> ./tools/build/src/user-config.jam
+  echo "using gcc : : ${PREFIX_DEPS}/bin/${CXX} ;" >> ./tools/build/src/user-config.jam
```
Comment (jeremyd2019) [inline mingw-w64-boost/PKGBUILD:167]: I'm not horribly fond of this, but just using ${CXX} didn't work.  Any better idea?

**Comment by MehdiChinoune** [comment]
Another idea is to patch `bootstrap.sh` to accept `--with-cxx=` option and pass it to `build.sh`

**Comment by jeremyd2019** [comment]
> Another idea is to patch `bootstrap.sh` to accept `--cxx=` option and pass it to `build.sh`

Yeah, it could also use that value in the toolset config in the project-config.jam it generates, so we wouldn't have to do it in user-config.jam

**Comment by MehdiChinoune** [comment]
proposed patch (not tested)
```diff
--- a/bootstrap.sh
+++ b/bootstrap.sh
@@ -10,6 +10,7 @@
 
 BJAM=""
 TOOLSET=""
+CXX_COMPILER=""
 BJAM_CONFIG=""
 BUILD=""
 PREFIX=/usr/local
@@ -131,6 +132,10 @@
       TOOLSET=`expr "x$option" : "x-*with-toolset=\(.*\)"`
       ;;
 
+    -with-cxx=* | --with-cxx=* )
+      CXX_COMPILER=`expr "x$option" : "x-*with-cxx=\(.*\)"`
+      ;;
+
     -*)
       { echo "error: unrecognized option: $option
 Try \`$0 --help' for more information." >&2
@@ -226,7 +231,11 @@
 if test "x$BJAM" = x; then
   $ECHO "Building B2 engine.."
   pwd=`pwd`
-  CXX= CXXFLAGS= "$my_dir/tools/build/src/engine/build.sh" ${TOOLSET}
+  if test "x$CXX_COMPILER" != x; then
+    CXX= CXXFLAGS= "$my_dir/tools/build/src/engine/build.sh" ${TOOLSET} --cxx=${CXX_COMPILER}
+  else
+    CXX= CXXFLAGS= "$my_dir/tools/build/src/engine/build.sh" ${TOOLSET}
+  fi
   if [ $? -ne 0 ]; then
       echo
       echo "Failed to build B2 build engine"
```

**Comment by jeremyd2019** [comment]
Maybe also
```diff
@@ -335,6 +344,10 @@

 # Generate user-config.jam
 echo "Generating B2 configuration in project-config.jam for $TOOLSET..."
+TOOLSET_OPTION=""
+if test "x$CXX_COMPILER" != x; then
+  TOOLSET_OPTION=" : : `which ${CXX_COMPILER}`"
+fi
 cat > project-config.jam <<EOF
 # B2 Configuration
 # Automatically generated by bootstrap.sh
@@ -347,7 +360,7 @@
 # file.
 if ! $TOOLSET in [ feature.values <toolset> ]
 {
-    using $TOOLSET ;
+    using $TOOLSET$TOOLSET_OPTION ;
 }

 project : default-build <toolset>$TOOLSET ;
```

**Comment by jeremyd2019** [comment]
> ```+  TOOLSET_OPTION=" : : `which ${CXX_COMPILER}`" ```

Actually this needs to be ```TOOLSET_OPTION=" : : `which ${CXX_COMPILER} | cygpath -mf -`"```

**Comment by jeremyd2019** [comment]
596773bfe3d4ae1e4aafd84cff5611f3f04be8b3 but not sure about putting `cygpath` in the script.  Building b2 as a separate step also seems to be less complicated/invasive.  But either way

**Comment by podsvirov** [comment]
And...

**Comment by podsvirov** [comment]
@jeremyd2019 , thanks for fixing the package rebuild issue!

**Pull request merged**